File: Arithmetic/1.twoSum/twoSum.php
<?php

//2792 ms	15.3 MB
function twoSum($nums, $target)
{
    $len = count($nums);
    for($i = 0; $i < $len - 1; $i ++){
        for($j = $i + 1; $j < $len; $j ++){
            if($nums[$i] + $nums[$j] == $target){
                return [$i, $j];
            }
            // 使用以下方法优化后运行时间减少到2164 ms
            // $temp = $target - $nums[$i];
            // if($nums[$j] == $temp){
            //     return [$i, $j];
            // }
        }
    }
    return [];
}

//36 ms	16.2 MB
function twoSum2($nums, $target) {
    $map = array_flip($nums);
    $len = count($nums);
    for($i = 0; $i < $len; $i++) 
    {
        $temp = $target-$nums[$i];
        if(!empty($map[$temp])) 
        {
            $j = $map[$temp];
            if($i != $j){
                return [$i, $j];
            }
        }
    }
    return [];
}

File: Arithmetic/001.TwoSum/TwoSum.java
import java.util.HashMap;
import java.util.Map;

public class TwoSum
{
    public static void main(String[] args)
    {
        int[] nums = {1, 2, 3, 4, 5};
        int target = 3;
        int[] res = twoSum2(nums, target);
        System.out.print(res[0]);
        System.out.print(res[1]);
    }

    //9 ms	38.4 MB
    public static int[] twoSum(int[] nums, int target)
    {
        Map<Integer, Integer> map = new HashMap<>();
        for(int i = 0; i < nums.length; i ++){
            map.put(nums[i], i);
        }

        for(int i = 0; i < nums.length; i ++){
            int targetNum = target - nums[i];
            if(map.containsKey(targetNum) && map.get(targetNum) != i){
                return new int []{i, map.get(targetNum)};
            }
        }

        throw new IllegalArgumentException("wrong");
    }

    //6 ms	38.1 MB
    public static int[] twoSum2(int[] nums, int target) {
        Map<Integer, Integer> map = new HashMap<>();
        for (int i = 0; i < nums.length; i++) {
            int complement = target - nums[i];
            if (map.containsKey(complement)) {
                return new int[] { map.get(complement), i };
            }
            map.put(nums[i], i);
        }
        throw new IllegalArgumentException("No two sum solution");
    }

}
